Client/Frontend/Login Management/LoginListViewController.swift

```diff
@@ -400,73 +418,91 @@ private class LoginCursorDataSource: NSObject, UITableViewDataSource {
     }
 
     @objc func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return loginsForSection(section).count
+        return loginsForSection(section)?.count ?? 0
     }
 
     @objc func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
         let cell = tableView.dequeueReusableCellWithIdentifier(LoginCellIdentifier, forIndexPath: indexPath) as! LoginTableViewCell
-
-        let login = loginAtIndexPath(indexPath)
+        let login = loginAtIndexPath(indexPath)!
         cell.style = .NoIconAndBothLabels
         cell.updateCellWithLogin(login)
-
         return cell
     }
 
     @objc func sectionIndexTitlesForTableView(tableView: UITableView) -> [String]? {
-        return sectionIndexTitles()
+        return titles.map { String($0) }
     }
 
     @objc func tableView(tableView: UITableView, sectionForSectionIndexTitle title: String, atIndex index: Int) -> Int {
-        guard let titles = sectionIndexTitles() where index < titles.count && index >= 0 else {
-            return 0
-        }
-        return titles.indexOf(title) ?? 0
+        return titles.indexOf(Character(title)) ?? 0
     }
 
     @objc func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
-        return sectionIndexTitles()?[section]
+        return String(titles[section])
     }
 
-    private func sectionIndexTitles() -> [String]? {
-        guard cursor?.count > 0 else {
-            return nil
+    private func computeLoginSections() {
+        titles.removeAll()
+        sections.removeAll()
+
+        guard allLogins.count > 0 else {
+            return
+        }
+
+        // Precompute the baseDomain, host, and hostname values for sorting later on. At the moment
+        // baseDomain() is a costly call because of the ETLD lookup tables.
+        var domainLookup = [GUID: (String?, String?, String)]()
```

Comment:
Maybe use named tuples here so that the code below is a bit more readable -- the magic numbers (i.e., `domainsA.2 < domainsB.2`) aren't very clear.